packages/lib/server/repository/team.ts

```diff
@@ -0,0 +1,149 @@
+  teamSelect = {
+    ...teamSelect,
+    metadata: true,
+  };
+  if (lookupBy.havingMemberWithId) where.members = { some: { userId: lookupBy.havingMemberWithId } };
+
+  if ("id" in lookupBy) {
+    where.id = lookupBy.id;
+  } else {
+    where.slug = lookupBy.slug;
+  }
+
+  if (isOrg) {
+    // We must fetch only the organization here.
+    // Note that an organization and a team that doesn't belong to an organization, both have parentId null
+    // If the organization has null slug(but requestedSlug is 'test') and the team also has slug 'test', we can't distinguish them without explicitly checking the metadata.isOrganization
+    // Note that, this isn't possible now to have same requestedSlug as the slug of a team not part of an organization. This is legacy teams handling mostly. But it is still safer to be sure that you are fetching an Organization only in case of isOrgView
+    where.metadata = {
+      path: ["isOrganization"],
+      equals: true,
+    };
+    // We must fetch only the team here.
+  } else {
+    if (forOrgWithSlug) {
+      where.parent = whereClauseForOrgWithSlugOrRequestedSlug(forOrgWithSlug);
+    }
+  }
+
+  log.debug({
+    orgSlug: forOrgWithSlug,
+    teamLookupBy: lookupBy,
+    isOrgView: isOrg,
+    where,
+  });
+
+  // teamSelect extends Prisma.TeamSelect but still teams doesn't contain a valid team as per TypeScript and thus it doesn't consider it having team.metadata, team.id and other fields
+  // This is the reason below code is using a lot of assertions.
+  const teams = await prisma.team.findMany({
+    where,
+    select: teamSelect,
+  });
+
+  const teamsWithParsedMetadata = teams
+    .map((team) => ({
+      ...team,
+      // Using Type assertion here because we know that the metadata is present and Prisma and TypeScript aren't playing well together
+      metadata: teamMetadataSchema.parse((team as { metadata: z.infer<typeof teamMetadataSchema> }).metadata),
+    }))
+    // In cases where there are many teams with the same slug, we need to find out the one and only one that matches our criteria
+    .filter((team) => {
+      console.log(team.metadata);
+      // We need an org if isOrgView otherwise we need a team
+      return isOrg ? team.metadata?.isOrganization : !team.metadata?.isOrganization;
+    });
+
+  if (teamsWithParsedMetadata.length > 1) {
+    log.error("Found more than one team/Org. We should be doing something wrong.", {
+      isOrgView: isOrg,
+      where,
+      teams: teamsWithParsedMetadata.map((team) => {
+        const t = team as unknown as { id: number; slug: string };
+        return {
+          id: t.id,
+          slug: t.slug,
+        };
+      }),
+    });
+  }
+
+  const team = teamsWithParsedMetadata[0];
+  if (!team) return null;
+  // HACK: I am not sure how to make Prisma in peace with TypeScript with this repository pattern
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return team as any;
```

Comment:
There is a return type on function already that should ensure that the function callers would get typesafey as long as this function is well tested.